### ISSUE TITLE
Specify a custom query class

### ISSUE THREAD

**Issue by immunda** [mitsuhiko/flask-sqlalchemy#319]
Devise a cohesive way to be able to provide a custom query class, which would be used for `Model.query_class`, `session` as `qry_class`, be set for `db.Query`, be used for `db.relationship` and `db.backref`. Essentially everywhere `BaseQuery` is used.

Thanks to the contributions and discussion from @justanr, @davidism & @RonnyPfannschmidt.

**Comment by RonnyPfannschmidt** [comment]
i dont see why #263 was not merged tho,

the next step would be to have a query_base parameter to __init__ i suppose,

#263 was about db.session.query using a different query class than flask-sqlalchemy to begin with

**Comment by immunda** [comment]
Because I would prefer to pass it as an arg to the constructor and have that be reflected throughout, rather than accept it as part of `sesssion_options` and then change the API a second time.

**Comment by justanr** [comment]
Doing this properly will require passing the custom query class all the way down to `_set_default_query_class`, including `_wrap_with_default_query_class` and `_include_sqlalchemy`. Otherwise some queries will have the new query class and some will have the `BaseQuery` class.

The other issue is by default `flask_sqlalchemy.Model` declares `query_class = BaseQuery` explicitly. The only *good* work around is doing something like this in `SQLAlchemy.__init__`:

```python
class SQLAlchemy(object):
     # may as well provide for custom base models as well
     # makes building something like SurrogatePKMixin easier
    def __init__(self, query_class=BaseQuery, model_class=Model, **others):
        self.Model = self.make_declarative_base(metadata, model_class)
        # not crazy about this part
        self.Model.query_class = query_class
        self.Query = query_class
```

**Comment by traut** [comment]
is that the last fix required for milestone 2.1?

**Comment by immunda** [comment]
Resolved in #328
